apf/apfSimplexAngleCalcs.cc

```diff
@@ -0,0 +1,366 @@
+#include "apf.h"
+#include "apfMesh.h"
+#include "apfShape.h"
+#include <iostream>
+#include <pcu_util.h>
+
+namespace apf {
+
+static Matrix3x3 getJacobianInMetric(const Matrix3x3& J, const Matrix3x3& Q)
+{
+  Matrix3x3 JT = transpose(J);
+  JT = transpose(Q)*JT;
+  return transpose(JT);
+}
+
+static Vector3 computeEdgeTangentAtVertex(Mesh* m, MeshEntity* edge,
+    MeshEntity* vert,
+    const Matrix3x3& Q)
+{
+  PCU_ALWAYS_ASSERT(m->getType(edge) == Mesh::EDGE);
+  PCU_ALWAYS_ASSERT(m->getType(vert) == Mesh::VERTEX);
+
+  // get the shared vertex
+  MeshEntity* ev[2];
+
+  m->getDownward(edge, 0, ev);
+
+  PCU_ALWAYS_ASSERT(ev[0] == vert || ev[1] == vert);
+
+  int index;
+  if (ev[0] == vert)
+    index = 0;
+  else
+    index = 1;
+
+  Vector3 xi;
+  if (index == 0)
+    xi = Vector3(-1.0, 0.0, 0.0);
+  else
+    xi = Vector3( 1.0, 0.0, 0.0);
+
+  MeshElement* edgeElem = createMeshElement(m, edge);
+  Matrix3x3 J;
+  getJacobian(edgeElem, xi, J);
+  destroyMeshElement(edgeElem);
+
+  J = getJacobianInMetric(J, Q);
+
+  Vector3 t;
+  if (index == 0)
+    t = Vector3(J[0]);
+  else
+    t = Vector3(0.0, 0.0, 0.0) - Vector3(J[0]);
+  return t.normalize();
+}
+
+
+Vector3 computeFaceNormalAtVertex(Mesh* m, MeshEntity* face,
+    MeshEntity* vert,
+    const Matrix3x3& Q)
+{
+  PCU_ALWAYS_ASSERT(m->getType(face) == Mesh::TRIANGLE);
```

Comment:
Is there some check in the higher level APIs that stops folks with mixed meshes from hitting this assert?